### PR TITLE
Add support for HCL

### LINKED ISSUE #1252: Support for Terraform
Support for Terraform: 

https://www.terraform.io/

### PR DESCRIPTION
**Pull request by outsideris**
Supporting [HCL(HashiCorp Configuration Language)](https://github.com/hashicorp/hcl).
Close #1252

### CODE REVIEW

**Comment by outsideris** [comment]
Sorry for late, I will fix them until next weekend.

**Comment by robmorgan** [comment]
@outsideris any luck? need a hand?

**Comment by outsideris** [comment]
I started to work on it. 
Sorry, I really was busy for company things.

File: components/prism-hcl.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+			lookbehind: true,
+			greedy: true
+		},
+		{
+			pattern: /#.*/
+		}
+	],
+	'keyword': [
+		{
+			pattern: /(resource|data)\s+(")?(\w+)(")?\s+(?=(")?([\w\-]+)(")?\s+{)/i,
```
[the remark this answers is not in the document]
Comment (outsideris) [reply components/prism-hcl.js]: 3.  Can I capture `non-capturing group` in `inside`?

File: components/prism-hcl.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+			inside: {
+				'type': [
+					{
+						pattern: /(resource|data)\s+(")?(\w+)(")?\s+/i,
```
[the remark this answers is not in the document]
Comment (outsideris) [reply components/prism-hcl.js]: 1. I don't understand it. I just matched it with above pattern.
2. I removed it.

File: components/prism-hcl.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+				'others': [
+					{
+						pattern: /[\w\*]+/i,
+						alias: 'string'
+					}
+				],
+			},
+		}
+	],
+	'string': /(["'])(\\(?:\r\n|[\s\S])|(?!\1)[^\\\r\n])*\1/,
```
[the remark this answers is not in the document]
Comment (outsideris) [reply components/prism-hcl.js]: 2. I think greedy is not needed here.

File: components/prism-hcl.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (RunDevelopment) [reply components/prism-hcl.js]: @outsideris If I understand you correctly you want something like #1472 (which is only an idea right now)?
I meant that `(")?(\w+)(")?` should be split into two part: First a simple `\w+` to match unquoted keys and second, one for quoted keys.

File: components/prism-hcl.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (RunDevelopment) [reply components/prism-hcl.js]: We usually don't want leading or trailing spaces inside a token.

File: components/prism-hcl.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (RunDevelopment) [reply components/prism-hcl.js]: 2. Comments alone can destroy your non-greedy strings: `"abc // what now?"`

File: components/prism-hcl.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	'boolean': /\b(?:true|false)\b/i,
+	'number': /-?\d+\.?\d*([kKmMgG]b?)?/,
```
[the remark this answers is not in the document]
Comment (outsideris) [reply components/prism-hcl.js]: [You are right](https://github.com/hashicorp/hcl#syntax).

> Numbers are assumed to be base 10. If you prefix a number with 0x, it is treated as a hexadecimal. If it is prefixed with 0, it is treated as an octal. Numbers can be in scientific notation: "1e10".

I have inspired it from sublime terraform addon, but I don't think `kKmMgG` is not needed here.

**Comment by outsideris** [comment]
I'm updating code as your great feedback. Thank you!!
I will push new commits soon after I'm done. Sorry for late response.

**Comment by robmorgan** [comment]
@outsideris no need to apologise for open source work mate ;)

File: components/prism-hcl.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (outsideris) [reply components/prism-hcl.js]: 1. HCL support only double-quoted strings.

2. It is hard to handle. HCL heavily use double-quoted strings like below. That's why I said greedy is not needed. If string is greedy, all interpolation in double-quoted is treated as string. Any idea?

[the remark this answers is not in the document]
Comment (RunDevelopment) [reply components/prism-hcl.js]: Just move `interpolation` into `['string'].inside`. I mean, isn't inpolation just part of a string?

This necessitates that the `string` pattern itself is aware of interpolation and thus it has to be adjusted accordingly.
Question: Does HCL support nested interpolation e.g. `"${file("/${var.trouble}.txt")}"` ?

[the remark this answers is not in the document]
Comment (outsideris) [reply components/prism-hcl.js]: > Just move interpolation into ['string'].inside. I mean, isn't inpolation just part of a string?

This is a good approach. I will try to find a way to support `interpolation` under `string`. I didn't fully understand prism's language definitions.

[the remark this answers is not in the document]
Comment (RunDevelopment) [reply components/prism-hcl.js]: Maybe a little help from my side: The `string` pattern should be:

```js
/"(?:[^\n\r\\$"]|\\.|\$(?:(?="|\$(?!\{))|[^\n\r\{$]|\$\{|\{(?:[^\n\r}"]|"(?:[^\n\r\\"]|\\.)*")*\}))*"/
```

It's more complex because of [this](https://www.terraform.io/docs/configuration/interpolation.html):
> You can escape interpolation with double dollar signs: `$${foo}` will be rendered as a literal `${foo}`.

**Comment by RunDevelopment** [comment]
@outsideris I'm sorry, I missed your last commit! I will the changes ASAP.

File: components/prism-hcl.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (outsideris) [reply components/prism-hcl.js]: I tried it. However, I feels it is too complex to maintain to me.

**Comment by outsideris** [comment]
@maartenvanderhoef It is not parse hcl2 and hcl2 is still [experimental](https://github.com/hashicorp/hcl2#experimental-hcl2).
For me, name "hcl" is fine because HashiCorp also call them as "HCL" and "HCL2". How do you think?

**Comment by outsideris** [comment]
I rebased onto latest master.

**Comment by maartenvanderhoef** [comment]
hi @outsideris, you're right, let's stick Hashicorp's own naming.

File: components/prism-hcl.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+		{
+			pattern: /[\w-\.]+(?=\s*=)/i,
+		},
+		{
+			pattern: /"[\w-\.]+"(?=\s*:)/i
+		}
+	],
+	'string': /"(?:[^\n\r\\"]|\\.)*"/,
+	'number': /(?:\d+e[+-]?\d+|0[0-7]+|0x[\da-f]+|\d+\.?\d*)/i,
+	'boolean': /\b(?:true|false)\b/i
```
[the remark this answers is not in the document]
Comment (outsideris) [reply components/prism-hcl.js]: https://github.com/hashicorp/hcl/blob/master/test-fixtures/basic_bool_string.hcl
I believe hcl treats `TRUE` as boolean.

File: components/prism-hcl.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+	'property': [
+		{
+			pattern: /[\w-\.]+(?=\s*=)/i,
+		},
+		{
+			pattern: /"[\w-\.]+"(?=\s*:)/i
+		}
+	],
+	'string': /"(?:[^\n\r\\"]|\\.)*"/,
+	'number': /(?:\d+e[+-]?\d+|0[0-7]+|0x[\da-f]+|\d+\.?\d*)/i,
```
[the remark this answers is not in the document]
Comment (outsideris) [reply components/prism-hcl.js]: fixed

File: components/prism-hcl.js [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+                ]
+			}
+		},
+		{
+			pattern: /([\w-]+)(?=\s+{)/
+		}
+	],
+	'interpolation': [
+		{
+			pattern: /\${([\w[\](+)\-*%^"',?=:/\s]+\.?)+}/i,
```
[the remark this answers is not in the document]
Comment (outsideris) [reply components/prism-hcl.js]: In my test, [given example](https://github.com/hashicorp/hcl/blob/master/test-fixtures/escape_backslash.hcl) isn't captured here. It is captured as `string`.
BTW, I fixed it to handle given example.

File: components/prism-hcl.js [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+						pattern: /[\w\*]+/,
+						alias: 'string'
+					}
+				],
+			},
+		}
+	],
+	'property': [
+		{
+			pattern: /[\w-\.]+(?=\s*=)/i,
```
[the remark this answers is not in the document]
Comment (outsideris) [reply components/prism-hcl.js]: fixed.

File: components/prism-hcl.js [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+						lookbehind: true,
+						alias: 'variable'
+					}
+				],
+				'keyword': /(terraform|var|self|count|module|path|data|local)/i,
+				'operator': /\$\{|[}.]/,
+				'function': [
+					{
+						pattern: /[\w]+(?=\()/,
+						alias: 'variable'
```
[the remark this answers is not in the document]
Comment (outsideris) [reply components/prism-hcl.js]: fixed.

File: components/prism-hcl.js [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+			inside: {
+				'type': [
+					{
+						pattern: /((terraform|var|self|count|module|path|data|local)\.)[\w\*]+/i,
+						lookbehind: true,
+						alias: 'variable'
+					}
+				],
+				'keyword': /(terraform|var|self|count|module|path|data|local)/i,
+				'operator': /\$\{|[}.]/,
```
[the remark this answers is not in the document]
Comment (outsideris) [reply components/prism-hcl.js]: fixed.

File: components/prism-hcl.js [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+				'function': [
+					{
+						pattern: /[\w]+(?=\()/,
```
[the remark this answers is not in the document]
Comment (outsideris) [reply components/prism-hcl.js]: fixed.

**Comment by outsideris** [comment]
I fixed as your feedback. Sorry for bordering you because I'm not regexp expert.

File: components/prism-hcl.js [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+					}
+				]
+			}
+		}
+	],
+	'property': [
+		/[^\\\r\n]+(?=\s*=)/,
+		/"[^\\\r\n]+"(?=\s*:)/
+	],
+	'string': /"(?:\\(?:\r\n|[\s\S])|(?!\1)[^\\\r\n])*"/,
```
Comment (RunDevelopment) [inline components/prism-hcl.js]: This pattern will also capture `"foo"bar"`.

Please use `/"/?:\\[\s\S]|[^\\"])*"/` to capture [multiline strings](https://github.com/hashicorp/hcl/blob/master/test-fixtures/multiline_literal.hcl).

File: components/prism-hcl.js [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+						pattern: /[\w\*]+/,
+						alias: 'string'
+					}
+				]
+			}
+		}
+	],
+	'property': [
+		/[^\\\r\n]+(?=\s*=)/,
+		/"[^\\\r\n]+"(?=\s*:)/
```
Comment (RunDevelopment) [inline components/prism-hcl.js]: Again, please use the proper string pattern instead of `"[^\\\r\n]"`.

File: components/prism-hcl.js [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+					{
+						pattern: /[\w\*]+/,
+						alias: 'string'
+					}
+				]
+			}
+		}
+	],
+	'property': [
+		/[^\\\r\n]+(?=\s*=)/,
```
Comment (RunDevelopment) [inline components/prism-hcl.js]: So `"" = true` and `/%$()[]{}ßöüü,.;:   = false` (including the spaces before `=`) are keys?

Why did you change the previous pattern? I thought `[\w-]+(?=\s*=)` worked or did it not?

File: components/prism-hcl.js [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+		/"[^\\\r\n]+"(?=\s*:)/
+	],
+	'string': /"(?:\\(?:\r\n|[\s\S])|(?!\1)[^\\\r\n])*"/,
+	'number': /(?:\d+.?\d*e[+-]?\d+|0[0-7]+|0x[\da-f]+|\d+\.?\d*)/i,
+	'boolean': /\b(?:true|false)\b/i,
+	'punctuation': /[=\[\]]/,
```
Comment (RunDevelopment) [inline components/prism-hcl.js]: Could you also include `{` and `}`?

File: components/prism-hcl.js [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+				]
+			}
+		}
+	],
+	'property': [
+		/[^\\\r\n]+(?=\s*=)/,
+		/"[^\\\r\n]+"(?=\s*:)/
+	],
+	'string': /"(?:\\(?:\r\n|[\s\S])|(?!\1)[^\\\r\n])*"/,
+	'number': /(?:\d+.?\d*e[+-]?\d+|0[0-7]+|0x[\da-f]+|\d+\.?\d*)/i,
```
Comment (RunDevelopment) [inline components/prism-hcl.js]: It has to be `\.?`. (Really easy to miss)

Also couldn't we make this `/\d+\.?\d*(?:e[+-]?\d+)?|0x[\da-f]+/i`? Octal numbers are included by the first `\d+`.

File: examples/prism-hcl.html [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+  ami           = "${data.aws_ami.ubuntu.id}"
+  instance_type = "t2.micro"
+
+  tags {
+    Name = "HelloWorld"
+  }
+}</code></pre>
+
+<h2>Provider</h2>
+<pre><code>resource "aws_instance" "web" {
```
Comment (RunDevelopment) [inline examples/prism-hcl.html]: Why did you use the same example twice?

File: tests/languages/hcl/string_feature.test [hunk 18]
```diff
@@ -0,0 +1,13 @@
+"double quoted\"' % string"
+'single quoted"\' % string'
```
Comment (RunDevelopment) [inline tests/languages/hcl/string_feature.test]: We don't support single quoted string, so you can remove this case.

**Comment by outsideris** [comment]
> Also, please replace all arrays which only contain one item with said item.

I believe I didn't understand what you mean exactly. I can't find more arrays which only contain one item.

File: components/prism-hcl.js (hunk 16 above)
Comment (RunDevelopment) [reply components/prism-hcl.js]: Yes, that's true. My bad.

**Comment by RunDevelopment** [comment]
Maybe another example: `hcl.keyword.inside.type` is currently defined as follows:

```js
'type': [ // start of array
	{	// one item
		pattern: /(resource|data|\s+)(?:[\w-]+|"[\w-]+")/i,
		lookbehind: true,
		alias: 'variable'
	}
] // end of array
```

But this can be replaced with a simpler expression:


```js
'type': {
	pattern: /(resource|data|\s+)(?:[\w-]+|"[\w-]+")/i,
	lookbehind: true,
	alias: 'variable'
}
```

Other examples like this can be found on line 7, 19, 32, 36, and 48.

File: components/prism-hcl.js [hunk 19]
```diff
@@ -55,11 +55,11 @@ Prism.languages.hcl = {
...
+		/[\w-\.]+(?=\s*=)/,
+		/"(?:\\[\s\S]|[^\\"])+"(?=\s*[:=])/,
 	],
-	'string': /"(?:\\(?:\r\n|[\s\S])|(?!\1)[^\\\r\n])*"/,
-	'number': /(?:\d+.?\d*e[+-]?\d+|0[0-7]+|0x[\da-f]+|\d+\.?\d*)/i,
-	'boolean': /\b(?:true|false)\b/i,
-	'punctuation': /[=\[\]]/,
+	'string': /"(?:\\[\s\S]|[^\\"])*"/,
+	'number': /0x[\da-f]+|\d+\.?\d*(?:e[+-]?\d+)?/i,
+	'boolean': /\b(true|false)\b/i,
```
Comment (RunDevelopment) [inline components/prism-hcl.js]: Why did make this a capturing group?

File: components/prism-hcl.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (RunDevelopment) [reply components/prism-hcl.js]: @outsideris 
I'm going to reuse this thread when talking about interpolation. I hope that's alright.

---

This pattern is a little simpler (and is based on the new string pattern): 
_(1)_

```js
/"(?:[^\\$"]|\\[\s\S]|\$(?:(?=")|\$+|[^"${])|\$\{(?:[^{}"]|"(?:[^\\"]|\\[\s\S])*")*\})*"/
```

I can think of simpler patterns but all of them will fail in O(2^n) in certain cases, meaning that they are practically useless.
So that is the simplest (useful) pattern I can think of which handles the following cases correctly:

```hcl
"foo\nbar" // simple

"$foo", "a$b", "foo$" // strings containing $
"${","}" // simple interpolation

// two strings in each line because the interpolation is escaped
// (GitHub gets all of these wrong)
"$${","}"
"$$${","}"
"$$$${","}"
"$$$$${","}"
```

We can then use the following pattern to get the interpolations parts of a string: 
_(2)_

```js
{
    pattern: /(^|[^$])\$\{(?:[^{}"]|"(?:[^\\"]|\\[\s\S])*")*\}/,
    lookbehind: true
}
```
(Everything after the lookbehind is copied from the above string pattern _(1)_.)

Using these two patterns we can make a string pattern similar to [JavaScript's template strings](https://github.com/PrismJS/prism/blob/master/components/prism-javascript.js#L60) using _(1)_ for `template-string` and _(2)_ for `interpolation`.
We have to adjust the inside of `interpolation` of course, but for that, we can use your current interpolation's inside.

**Comment by outsideris** [comment]
fixed.

**Comment by RunDevelopment** [comment]
@outsideris 
Good!
The things left to do now are:

1. Interpolation as discussed [here](https://github.com/PrismJS/prism/pull/1594#discussion_r242251517).
2. [Here-doc strings](https://github.com/hashicorp/hcl/blob/master/test-fixtures/multiline.hcl).

**Comment by outsideris** [comment]
@RunDevelopment 
I moved interpolation into `string` as your advice. Thanks. -> 8022bed85c91b886d6399e6187fd06b0b8a6edf0
I supported heredoc. -> c05d49806eaeedd02e92ca18a4e87cbb7a910de7

File: components/prism-hcl.js [hunk 20]
```diff
@@ -0,0 +1,60 @@
+Prism.languages.hcl = {
+	'comment': /(?:\/\/|#).*|\/\*[\s\S]*?(?:\*\/|$)/,
+	'heredoc': {
+		pattern: /<<(\w+)(?:[^\1])*?\1/,
```
[the remark this answers is not in the document]
Comment (outsideris) [reply components/prism-hcl.js]: Really thanks. I changed `/<<-?(\w+)[\s\S]*?^\s*<<\1/m` to `/<<-?(\w+)[\s\S]*?^\s*\1/m`, because last `<<` is not needed.

File: components/prism-hcl.js [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+		{
+			pattern: /[\w-]+(?=\s+{)/
+		}
+	],
+	'property': [
+		/[\w-\.]+(?=\s*=(?!=))/,
+		/"(?:\\[\s\S]|[^\\"])+"(?=\s*[:=])/,
+	],
+	'string': {
+		pattern: /"(?:[^\\$"]|\\[\s\S]|\$(?:(?=")|\$+|[^"${])|\$\{(?:[^{}"]|"(?:[^\\"]|\\[\s\S])*")*\})*"/,
```
[the remark this answers is not in the document]
Comment (outsideris) [reply components/prism-hcl.js]: ok

File: components/prism-hcl.js [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+					'type': {
+						pattern: /((?:terraform|var|self|count|module|path|data|local)\.)[\w\*]+/i,
+						lookbehind: true,
+						alias: 'variable'
+					},
+					'keyword': /terraform|var|self|count|module|path|data|local/ig,
+					'function': /\w+(?=\()/,
+					'string': /"(?:\\[\s\S]|[^\\"])*"/,
+					'punctuation': /[!\$#%&'()*+,.\/;<=>@\[\\\]^`{|}~?:]/,
+					'number': /-?\d+\.?\d*/,
```
[the remark this answers is not in the document]
Comment (outsideris) [reply components/prism-hcl.js]: fixed

File: components/prism-hcl.js [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+			'interpolation': {
+				pattern: /(^|[^$])\$\{(?:[^{}"]|"(?:[^\\"]|\\[\s\S])*")*\}/,
+				lookbehind: true,
+				inside: {
+					'type': {
+						pattern: /((?:terraform|var|self|count|module|path|data|local)\.)[\w\*]+/i,
+						lookbehind: true,
+						alias: 'variable'
+					},
+					'keyword': /terraform|var|self|count|module|path|data|local/ig,
```
[the remark this answers is not in the document]
Comment (outsideris) [reply components/prism-hcl.js]: fixed

File: components/prism-hcl.js [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+					'function': /\w+(?=\()/,
+					'string': /"(?:\\[\s\S]|[^\\"])*"/,
```
[the remark this answers is not in the document]
Comment (outsideris) [reply components/prism-hcl.js]: ok

File: components/prism-hcl.js [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+					'punctuation': /[!\$#%&'()*+,.\/;<=>@\[\\\]^`{|}~?:]/,
```
[the remark this answers is not in the document]
Comment (outsideris) [reply components/prism-hcl.js]: Good catch.

File: components/prism-hcl.js (hunk 23 above)
[the remark this answers is not in the document]
Comment (outsideris) [reply components/prism-hcl.js]: fixed

File: components/prism-hcl.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (RunDevelopment) [reply components/prism-hcl.js]: My bad.

**Comment by outsideris** [comment]
Fixed! :)

File: components/prism-hcl.js [hunk 26]
```diff
@@ -1,25 +1,26 @@
...
 	'heredoc': {
 		pattern: /<<-?(\w+)[\s\S]*?^\s*\1/m,
+		greedy: true,
 		alias: 'string'
 	},
-	'comment': /(?:\/\/|#).*|\/\*[\s\S]*?(?:\*\/|$)/,
 	'keyword': [
 		{
-			pattern: /(?:resource|data)\s+(?:[\w-]+|"[\w-]+")(?=\s+"[\w-]+"\s+{)/i,
+			pattern: /(?:resource|data)\s+(?:[\w-]+|"(?:\\[\s\S]|[^\\"])*")(?=\s+"[\w-]+"\s+{)/i,
```
Comment (RunDevelopment) [inline components/prism-hcl.js]: The lookahead still contains the old `"[\w-]+"`.

File: components/prism-hcl.js [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+						lookbehind: true,
+						alias: 'variable'
+					},
+					'keyword': /\b(?:terraform|var|self|count|module|path|data|local)\b/i,
+					'function': /\w+(?=\()/,
+					'string': {
+						pattern: /"(?:\\[\s\S]|[^\\"])*"/,
+						greedy: true,
+					},
+					'number': /0x[\da-f]+|\d+\.?\d*(?:e[+-]?\d+)?/i,
```
Comment (RunDevelopment) [inline components/prism-hcl.js]: I tested a little and names containing numbers are incorrectly highlighted as `numbers`.
Please add make it `/\b0x.../i` to fix this.

File: components/prism-hcl.js (hunk 26 above)
Comment (outsideris) [reply components/prism-hcl.js]: Sorry, I missed them.

**Comment by outsideris** [comment]
Done.

**Comment by outsideris** [comment]
@RunDevelopment Fixed. Sorry for I missed it.

**Comment by outsideris** [comment]
@RunDevelopment You are awesome. I couldn't make this PR without you. Thank you.
